cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/Internal.hs

```diff
@@ -0,0 +1,883 @@
+  p2 <- mallocForeignPtrBytes sizeFr
+  x <- withForeignPtr p2 (go . FrPtr)
+  return (x, Fr p2)
+
+withNewFr_ :: (FrPtr -> IO a) -> IO a
+withNewFr_ = fmap fst . withNewFr
+
+withNewFr' :: (FrPtr -> IO a) -> IO Fr
+withNewFr' = fmap snd . withNewFr
+
+cloneFr :: Fr -> IO Fr
+cloneFr (Fr a) = do
+  b <- mallocForeignPtrBytes sizeFr
+  withForeignPtr a $ \ap ->
+    withForeignPtr b $ \bp ->
+      copyBytes bp ap sizeFr
+  return (Fr b)
+
+scalarToInteger :: Scalar -> IO Integer
+scalarToInteger scalar = withScalar scalar $ \scalarPtr -> do
+  allocaBytes sizeScalar $ \rawPtr -> do
+    c_blst_bendian_from_scalar rawPtr scalarPtr
+    cstrToInteger rawPtr sizeScalar
+
+cstrToInteger :: Ptr CChar -> Int -> IO Integer
+cstrToInteger p l = do
+  go l (castPtr p)
+  where
+    go :: Int -> Ptr CUChar -> IO Integer
+    go 0 _ = return 0
+    go n ptr = do
+      val <- peek ptr
+      res <- go (pred n) (plusPtr ptr 1)
+      return $ res .|. shiftL (fromIntegral val) (8 * pred n)
+
+integerToBS :: Integer -> ByteString
+integerToBS 0 = BS.empty
+integerToBS n
+  | n < 0
+  = error "Cannot convert negative Integer to ByteString"
+  | otherwise
+  = BS.snoc
+      (integerToBS (n `shiftR` 8))
+      (fromIntegral n)
+
+padBS :: Int -> ByteString -> ByteString
+padBS i b
+  | i > BS.length b
+  = BS.replicate (i - BS.length b) 0 <> b
+  | otherwise
+  = b
+
+integerAsCStr :: Integer -> (Ptr CChar -> Int -> IO a) -> IO a
+integerAsCStr n f = do
+  let bs = integerToBS n
+  BS.useAsCStringLen bs $ uncurry f
+
+integerAsCStrL :: Int -> Integer -> (Ptr CChar -> Int -> IO a) -> IO a
+integerAsCStrL i n f = do
+  let bs = padBS i $ integerToBS n
+  BS.useAsCStringLen bs $ uncurry f
+
+scalarFromInteger :: Integer -> IO Scalar
+scalarFromInteger n = do
+  withNewScalar' $ \scalarPtr -> do
+    integerAsCStrL sizeScalar (n `mod` scalarPeriod) $ \str _length -> do
+      c_blst_scalar_from_bendian scalarPtr str
+
+---- Unsafe types
+
+newtype ScalarPtr = ScalarPtr (Ptr Void)
+newtype FrPtr = FrPtr (Ptr Void)
+
+---- Raw Scalar / Fr functions
+
+foreign import ccall "size_blst_scalar" c_size_blst_scalar :: CSize
+foreign import ccall "size_blst_fr" c_size_blst_fr :: CSize
+
+foreign import ccall "blst_scalar_fr_check" c_blst_scalar_fr_check :: ScalarPtr -> IO Bool
+
+foreign import ccall "blst_scalar_from_fr" c_blst_scalar_from_fr :: ScalarPtr -> FrPtr -> IO ()
+foreign import ccall "blst_fr_from_scalar" c_blst_fr_from_scalar :: FrPtr -> ScalarPtr -> IO ()
+foreign import ccall "blst_scalar_from_be_bytes" c_blst_scalar_from_be_bytes :: ScalarPtr -> Ptr CChar -> CSize -> IO Bool
+foreign import ccall "blst_scalar_from_bendian" c_blst_scalar_from_bendian :: ScalarPtr -> Ptr CChar -> IO ()
+
+---- Raw P1 functions
+
+foreign import ccall "size_blst_p1" c_size_blst_p1 :: CSize
+foreign import ccall "blst_p1_on_curve" c_blst_p1_on_curve :: P1Ptr -> IO Bool
+
+foreign import ccall "blst_p1_add_or_double" c_blst_p1_add :: P1Ptr -> P1Ptr -> P1Ptr -> IO ()
```

Comment:
Also rename `c_blst_p1_add` to `c_blst_p1_add_or_double`

Reply:
Done.